[WIP] Add event for custom entry points

runelaenen opened the pull request:
**WORK IN PROGRESS** but constructive feedback is always welcome

### 1. Why is this change necessary?
To allow themes to add more navigational entry points.

### 2. What does this change do, exactly?
Add an event that will collect all entry point ID's instead of them being 'hardcoded' all over the core.

### 3. Describe each step to reproduce the issue or behaviour.
WIP

### 4. Please link to the relevant issues (if any).


### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/master/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

Commit:
Add event for custom entry points

mitelg commented on src/Core/Content/Category/Event/SalesChannelEntryPointsEvent.php:
```diff
@@ -0,0 +1,39 @@
+
+    public function __construct(array $ids, SalesChannelEntity $salesChannelEntity)
+    {
+        $this->ids = $ids;
```
what does this array contain? I think it would be useful do add some PHPStan/Psalm like Annotations or at least some PHPdoc comment to explain. Also `ids` is not quite descriptive :blush:

mitelg commented on src/Core/Content/Category/Event/SalesChannelEntryPointsEvent.php:
```diff
@@ -0,0 +1,39 @@
+     */
+    protected $salesChannelEntity;
+
+    public function __construct(array $ids, SalesChannelEntity $salesChannelEntity)
```
is it possible to add the whole SalesChannelContext here?

runelaenen replied:
Not really, as the event is also used in places where the saleschannelcontext is not readily available.

shyim commented on src/Core/Content/Category/SalesChannel/NavigationRoute.php:
```diff
@@ -19,6 +20,7 @@
 use Shopware\Core\System\SalesChannel\Entity\SalesChannelRepositoryInterface;
 use Shopware\Core\System\SalesChannel\SalesChannelContext;
 use Shopware\Core\System\SalesChannel\SalesChannelEntity;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
```
```suggestion
use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
```

shyim commented on src/Core/Content/Category/SalesChannel/NavigationRoute.php:
```diff
@@ -47,16 +49,28 @@ class NavigationRoute extends AbstractNavigationRoute
+    /**
+     * @var array
+     */
+    private $entryPointIds;
```
Why we need this state in the class?
This makes unit testing hard

shyim commented on src/Core/Content/Category/Service/CategoryBreadcrumbBuilder.php:
```diff
@@ -20,12 +22,20 @@ class CategoryBreadcrumbBuilder
-    {
+    public function __construct(
+        ?EntityRepositoryInterface $categoryRepository = null,
+        EventDispatcherInterface $dispatcher
```
This is not allowed in PHP 8. 
https://php.watch/versions/8.0/deprecate-required-param-after-optional

shyim commented:
@runelaenen You should use Draft PRs instead marking them by title WIP 😊

runelaenen commented:
> @runelaenen You should use Draft PRs instead marking them by title WIP 😊

Ah I didn't know that that was a thing 😂 I will do that the next time!

mitelg commented on src/Core/Content/Category/Event/SalesChannelEntryPointsEvent.php:
```diff
@@ -0,0 +1,52 @@
+
+use Shopware\Core\System\SalesChannel\SalesChannelEntity;
+
+class SalesChannelEntryPointsEvent
```
hey @runelaenen 
yesterday a new ADR was accepted and merged: 
https://github.com/shopware/platform/commit/dacf9cf492ba602f0966d422d381c36acb7fe79f

> From now on every new event must implement the `Shopware\Core\Framework\Event\ShopwareEvent` interface. If a `Shopware\Core\System\SalesChannel\SalesChannelContext` is also available, the `Shopware\Core\Framework\Event\ShopwareSalesChannelEvent` interface must be implemented instead.
Some existing events were adjusted, which led to some deprecations.

would be nice if you could consider this new convention :blush: :+1:

Commit:
Merge branch 'master' of github.com:shopware/platform into feature/custom-entry-points

Commit:
Implement ShopwareEvent

Commit:
Fix phpcs and phpunit

runelaenen commented:
Right now I've implemented the functionality using an Event. However, in theory, an event is immutable. (I know that Shopware does mutate events in places, as it's why I've implemented it here in this way too) 
Would it be better to introduce a new DI tag and Interface with which you can register entry points, instead of an event with setters?

@shyim @mitelg What do you think?

mitelg commented:
hey @runelaenen 
the way with the event is okay for us :+1: 

but we asked ourselves if it might be possible, to centralise the dispatch of the event. so that there is one point where the entrypoints are built. maybe you could do that, when the SalesChannelContext is built and the entry points are a new property of the context. what do you think about that?

runelaenen commented:
Hi @mitelg 
Sounds good in theory, but that would mean that the entrypoints cannot easily be loaded without a SalesChannelContext. What should happen in console commands, ... where no SalesChannelContext is available? 
Of course it would still be possible to dispatch the event there too with just the SalesChannelEntity object.

mitelg commented:
this is just an idea :blush: 
but we thought it would be easier to understand if this event is dispatched in one place and not in three or four places.

runelaenen commented:
@mitelg  Well yes, I agree that it would be nice to have the entrypoints readily available in the SalesChannelContext object.

I did however just find another issue. https://github.com/shopware/platform/pull/1486/checks?check_run_id=1443053119
Neither DI nor the SalesChannelEntity is available in the CategoryEntity. How should I handle this best?

mitelg commented:
yeah, that is another issue, that there is such business logic in an entity :see_no_evil:

runelaenen commented:
> yeah, that is another issue, that there is such business logic in an entity 🙈

The code has been deprecated, but that does not really help at this moment of course :/

philipreinken commented:
Hi @runelaenen , any news here? 🙂

runelaenen commented:
Hi @philipgatzka 

I'm closing this PR and re-doing it once 6.4 is released so the breadcrumb logic is removed from the CategoryEntity class.
